### PR TITLE
environmentd: Add boostrap role to startup

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds the concept of a bootstrap role to startup. If a bootstrap role is specified when starting Materialize, and it's the first time that this instance of Materialize has been started, then a bootstrap role will be created with the given name that has elevated privileges. The role will have the `CREATEDB` and `CREATECLUSTER` attribute, and the `CREATE` privilege on the `materialize` database, the `materialize.public` schema, and the `default` cluster. The purpose of this role is to make testing and local development easier when RBAC is enabled.

### Motivation
This PR adds a feature that has not yet been specified.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/catalog/storage.rs [hunk 1]
```diff
@@ -420,6 +420,92 @@ async fn migrate(
...
+                let audit_log_id = txn.get_and_increment_id(AUDIT_LOG_ID_ALLOC_KEY.to_string())?;
+                txn.audit_log_updates.push((
+                    AuditLogKey {
+                        event: VersionedEvent::new(
+                            audit_log_id,
+                            EventType::Create,
+                            ObjectType::Role,
+                            EventDetails::IdNameV1(mz_audit_log::IdNameV1 {
+                                id: role_id.to_string(),
+                                name: "materialize".into(),
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/adapter/src/catalog/storage.rs]: Yes, just updated.

**Review by mjibson — APPROVED** [review]
LGTM. This wasn't a problem before because rbac isn't fully enabled yet?

**Comment by jkosh44** [comment]
> LGTM. This wasn't a problem before because rbac isn't fully enabled yet?

Yes, it still isn't a problem because we're going to leave RBAC disabled for awhile. I just want to get ahead of the issue.

File: src/environmentd/src/lib.rs [hunk 2]
```diff
@@ -128,6 +128,8 @@ pub struct Config {
     // === Special modes. ===
     /// Whether to permit usage of unsafe features.
     pub unsafe_mode: bool,
+    /// What role, if any, should be initially created with elevated privileges.
+    pub bootstrap_role: Option<String>,
```
Comment (benesch) [inline src/environmentd/src/lib.rs:132]: Hmm, this isn't really a "special mode." I'd move this down next to `bootstrap_system_parameters`. (Maybe that section should be renamed from "cloud options" to .. something else, not sure exactly what.)

**Review by benesch — COMMENTED** [review]
(Heads up: didn't review the code in storage.rs. Figured other folks were on that.)

**Comment by jkosh44** [comment]
> Just for clarification: Is this only for testing and not for productive use? If so, should we make the cmdline parameter make that clearer?

Yes, it's only for testing. Do you mean the name of the cmdline parameter or the docs for the cmdline paramete?

**Comment by def-** [comment]
Name

File: src/environmentd/src/lib.rs (hunk 2 above)
Comment (jkosh44) [reply src/environmentd/src/lib.rs:132]: What about moving it to the "Testing Options" section?

File: misc/python/materialize/mzcompose/services.py [hunk 3]
```diff
@@ -148,6 +148,8 @@ def __init__(
             "--orchestrator-process-prometheus-service-discovery-directory=/mzdata/prometheus",
         ]
 
+        environment += ["BOOTSTRAP_ROLE=materialize"]
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply misc/python/materialize/mzcompose/services.py]: Done

**Comment by jkosh44** [comment]
> Name

Done, let me know what you think of the name.

**Comment by benesch** [comment]
Hmm, I liked the original name better! I am generally not a fan of command line parameters that prescribe their use in their name. This flag has clearly described semantics, and I think it should be named after those semantics. AFAICT, there's nothing inherently scary if the cloud infrastructure one day decided to pass `--bootstrap-role=name.of.first.user.on.account@wherever.com`. I'd also prefer to stick with the `bootstrap` prefix, as we use that in several other command line parameters (`--bootstrap-system-parameter`, for example). 

If we're worried that "bootstrap role" is not descriptive enough (though I think it is), then perhaps we could move in the direction of further describing the semantics, e.g. `--bootstrap-default-role` or `--bootstrap-admin-role`.

**Comment by jkosh44** [comment]
> Hmm, I liked the original name better! I am generally not a fan of command line parameters that prescribe their use in their name. This flag has clearly described semantics, and I think it should be named after those semantics. AFAICT, there's nothing inherently scary if the cloud infrastructure one day decided to pass `--bootstrap-role=name.of.first.user.on.account@wherever.com`. I'd also prefer to stick with the `bootstrap` prefix, as we use that in several other command line parameters (`--bootstrap-system-parameter`, for example).
> 
> If we're worried that "bootstrap role" is not descriptive enough (though I think it is), then perhaps we could move in the direction of further describing the semantics, e.g. `--bootstrap-default-role` or `--bootstrap-admin-role`.

I also liked the old name better tbh, so I switched it back. I think `--bootstrap-role` is descriptive enough.

In terms of renaming the cloud section, does "Bootstrap options" make sense?

**Comment by benesch** [comment]
> In terms of renaming the cloud section, does "Bootstrap options" make sense?

Yeah, SGTM.

**Pull request merged**